### DISCUSSION TITLE
Set default authentication provider

### DISCUSSION THREAD

**Discussion by dorianim** [directus/directus#9298] [Feature Requests]
Hi,

Is it possible to set the default authentication provider to anything other than local? I tried overriding the DEFAULT provider, but this seems not to be allowed ( https://github.com/directus/directus/blob/b7ee318e4329f3623c41b222b6ba85433b8714ca/api/src/auth.ts#L41 )...

Is there any other way to select ldap instead of local by default?

Best Regards
Dorian

**Comment by aidenfoxx** [comment]
Not at the moment, but it's certainly possible if you fancy making a PR for it!

**Reply by dorianim** [reply]
Ok. So how would this best be implemented? Overriding default would be dangerous, as it would break login for the default admin. 
Maybe by changing the order of the items somewhere around here: https://github.com/directus/directus/blob/ded3d4b1412a4da711dfaa31e72a4fd4e7e31acc/app/src/routes/login/login.vue#L77
?

**Reply by dorianim** [reply]
I think, I could add an attribute `default` to the [`/auth` endpoint,](https://github.com/directus/directus/blob/ded3d4b1412a4da711dfaa31e72a4fd4e7e31acc/api/src/controllers/auth.ts#L191) which will then be used as the default provider by the app. This could be set by a new config variable like `AUTH_PROVIDERS_DEFAULT`.
Would this be feasible?

**Reply by aidenfoxx** [reply]
There shouldn't be a need to override anything. We have an API that the frontend calls that tells the frontend how to render the login. If the driver for "default" is LDAP, it should just render the `ldap-form` with no dramas.

The only thing that's uncertain is how to handle if a user overrides the default provider with SSO, and there are no other non-SSO flows. I guess in that instance we just hide the login form and just show the SSO links @rijkvanzanten?

**Reply by aidenfoxx** [reply]
I built the auth flow with the intention that we'd allow users to override the default provider, so honestly I think most stuff should just work if you remove the restriction:

https://github.com/directus/directus/blob/b7ee318e4329f3623c41b222b6ba85433b8714ca/api/src/auth.ts#L41

The only thing holding it back was the old OAuth flow, which is dead at this point.

**Reply by dorianim** [reply]
> honestly I think most stuff should just work if you remove the restriction

But default local admin won't work anymore, or am I missing something there?

**Reply by rijkvanzanten** [reply]
> The only thing that's uncertain is how to handle if a user overrides the default provider with SSO, and there are no other non-SSO flows. I guess in that instance we just hide the login form and just show the SSO links @rijkvanzanten?

I think we should have an optional "AUTH_DEFAULT_PROVIDER" or something along those lines which you can point to a configured auth provider. At that point, the app can simply hide the regular login form in favor of what it would already render for the other configured auth provider, seeing that that now exists in the array of configured providers 🤔

**Reply by aidenfoxx** [reply]
Aha, there is a little more to it. Right now the "default" provider is hard wired. We'd need to move it out to the `/auth` endpoint and as you say.

We'd also have to fix the auth controller since right now the "localAuthRouter" gets bound to a special endpoint. We'd have to make sure that the new default also binds to that special endpoint.

**Reply by rijkvanzanten** [reply]
Totally! I didn't include any of the required API changes in my comment 😁

**Reply by aidenfoxx** [reply]
@rijkvanzanten I guess you're worried about users accidentally overriding the default auth provider? We could make it clear in the docs that setting a provider with the name "default" might screw up their system. 😋

**Reply by rijkvanzanten** [reply]
I'm wondering whether we do it based on the auth provider name "default", or on a dedicated flag 🤔 To me it feels a little cleaner to have a dedicated env var that says "Use this configured auth provider as the default auth provider" regardless of what it's name is 🤔

**Reply by aidenfoxx** [reply]
Really? I think it's cleaner to have it based on provider name. Will we ever agree on anything? 😂

**Reply by rijkvanzanten** [reply]
Haha nope! 

There's also a third option: Not allowing to override the default login, but just adding a flag to _disable_ it. That would make implementation a lot easier, as we won't have to do any remapping of that root /login endpoint. If you'd disable the default login and add a single custom one, the end result would be the same

**Reply by aidenfoxx** [reply]
Sounds good to me!

**Reply by rijkvanzanten** [reply]
> Will we ever agree on anything? 😂

We found our answer! _Yes_ 😄

**Reply by dorianim** [reply]
Will one of you implement this or should I try to do it?

**Reply by aidenfoxx** [reply]
I won't be looking into it and Rijk is busy as hell. I think you're free to give it a go.

**Reply by LasseRosenow** [reply]
But now the issue is, I for example would still like to be able to login with my admin account which is not in LDAP.

For me the ideal solution would be, that "LDAP" would just be selected by default by that I could click on the drop down and switch to "Default" whenever I want

**Reply by aidenfoxx** [reply]
@LasseRosenow You technically can do that. Just disable the default login and add a different "local" auth driver in your config. Something like:

```
AUTH_DISABLE_DEFAULT=true
AUTH_PROVIDERS="ldap,admin"

<your_ldap_config>

AUTH_ADMIN_DRIVER="local"
```

That way LDAP should be above your "admin" login provider. It's kinda ugly, but it should work.

**Reply by LasseRosenow** [reply]
aaah thats a nice hack, thank you very much :)

**Reply by miconx** [reply]
if i do this then the new "admin" provider doesnt show the login form anymore 
did i miss something? 
![image](https://user-images.githubusercontent.com/504850/201108630-ecea2337-c227-4c7f-bbd3-8b7a01be8c49.png)

**Reply by fynrup** [reply]
I have the same problem as @miconx. @aidenfoxx can you explain why the login form is not shown any longer?

**Reply by aidenfoxx** [reply]
Please create an issue so the problem can be properly investigated. :+1:

**Reply by fynrup** [reply]
Linked issue: #17340
